src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -109,21 +137,41 @@ public WriteAheadJournal(StorageEnvironment env)
         public ImmutableAppendOnlyList<JournalFile> Files => _files;
 
         public JournalApplicator Applicator => _journalApplicator;
+        
+        public bool HasBranchCommits => _mergedCommitsQueue.IsEmpty is false;
 
-        private JournalFile NextFile(int numberOf4Kbs = 1)
+        private JournalFile NextFile(long numberOf4Kbs, int numberOfRegisteredEnvironments)
         {
             var now = DateTime.UtcNow;
+            
+            numberOfRegisteredEnvironments = Math.Max(numberOfRegisteredEnvironments,
+                // if we have a decrease in the number of registered environments when creating a new
+                // file from the previous one, we'll average the two values to ensure more graceful 
+                // decay of journal file size
+                (_currentNumberOfRegisteredEnvironments + numberOfRegisteredEnvironments) / 2);
+
+            long maxLogFileSize = _env.Options.MaxLogFileSize;
+            if (numberOfRegisteredEnvironments > 1)
+            {
+                // If the previous log had multiple registered envs (root/branch model)
+                // we'll increase the max file size of the journals to accomodate that. With
+                // the more environments we have, the bigger the file is going to be.
+                maxLogFileSize *= (int)Math.Ceiling(Math.Log2(numberOfRegisteredEnvironments));
+            } 
+
+            _currentNumberOfRegisteredEnvironments = numberOfRegisteredEnvironments;    
```

Comment:
is this a concern here that when this method is accessed concurrently then we might be overriding that?